src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3StreamTests.cs

```diff
@@ -2101,5 +2102,261 @@ public async Task HEADERS_Received_HeaderBlockContainsTEHeader_ValueIsTrailers_N
+            var headers = new[]
+            {
+                new KeyValuePair<string, string>(HeaderNames.Method, "POST"),
+                new KeyValuePair<string, string>(HeaderNames.Path, "/"),
+                new KeyValuePair<string, string>(HeaderNames.Scheme, "http"),
+                new KeyValuePair<string, string>(HeaderNames.ContentLength, "12"),
+            };
+            var requestStream = await InitializeConnectionAndStreamsAsync(async context =>
+            {
+                var buffer = new byte[100];
+                var read = await context.Request.Body.ReadAsync(buffer, 0, buffer.Length);
+                Assert.Equal(12, read);
+                read = await context.Request.Body.ReadAsync(buffer, 0, buffer.Length);
+                Assert.Equal(0, read);
+            });
+
+            await requestStream.SendHeadersAsync(headers, endStream: false);
+            await requestStream.SendDataAsync(new byte[12], endStream: true);
+
+            var receivedHeaders = await requestStream.ExpectHeadersAsync();
+
+            await requestStream.ExpectReceiveEndOfStream();
+
+            Assert.Equal(3, receivedHeaders.Count);
+            Assert.Contains("date", receivedHeaders.Keys, StringComparer.OrdinalIgnoreCase);
+            Assert.Equal("200", receivedHeaders[HeaderNames.Status]);
+            Assert.Equal("0", receivedHeaders[HeaderNames.ContentLength]);
+        }
+
+        [Fact]
+        public async Task MaxRequestBodySize_ContentLengthOver_413()
+        {
+#pragma warning disable CS0618 // Type or member is obsolete
+            BadHttpRequestException exception = null;
+#pragma warning restore CS0618 // Type or member is obsolete
+            _serviceContext.ServerOptions.Limits.MaxRequestBodySize = 10;
+            var headers = new[]
+            {
+                new KeyValuePair<string, string>(HeaderNames.Method, "POST"),
+                new KeyValuePair<string, string>(HeaderNames.Path, "/"),
+                new KeyValuePair<string, string>(HeaderNames.Scheme, "http"),
+                new KeyValuePair<string, string>(HeaderNames.ContentLength, "12"),
+            };
+            var requestStream = await InitializeConnectionAndStreamsAsync(async context =>
+            {
+#pragma warning disable CS0618 // Type or member is obsolete
+                exception = await Assert.ThrowsAsync<BadHttpRequestException>(async () =>
+#pragma warning restore CS0618 // Type or member is obsolete
+                {
+                    var buffer = new byte[100];
+                    while (await context.Request.Body.ReadAsync(buffer, 0, buffer.Length) > 0) { }
+                });
+                ExceptionDispatchInfo.Capture(exception).Throw();
+            });
+
+            await requestStream.SendHeadersAsync(headers, endStream: false);
+
+            var receivedHeaders = await requestStream.ExpectHeadersAsync();
+
+            await requestStream.ExpectReceiveEndOfStream();
```

Comment:
Should there be a stream reset received here?

Reply:
https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-4.1-14
> A server can send a complete response prior to the client sending an entire request if the response does not depend on any portion of the request that has not been sent and received. When the server does not need to receive the remainder of the request, it MAY abort reading the request stream, send a complete response, and cleanly close the sending part of the stream. The error code H3_NO_ERROR SHOULD be used when requesting that the client stop sending on the request stream.

Reply:
Yes. This is the follow up - https://github.com/dotnet/aspnetcore/issues/31970